### PR TITLE
ENG-1070: Alias `contract` to `collection`

### PR DESCRIPTION
**Pull request by timmyjose**
Changes:
  - added alias `contract` for `collection` in parser-generator grammar.
  - added parser test for `contract`.
  - added alias `contract` for `collection` in lexer + test.
  - added test in `polylang` tests.

### CODE REVIEW

**Comment by timmyjose** [comment]
The latest commit has the following semantic changes:

  * Makes `contract` first-class instead of `collection`. `collection` is now an alias for `contract`.
  * `Polylang` internal structures renamed from `collection` names to `contract` names.
  * `Polylang` grammar updated to rename `collection` types/structures to `contract.
  * `stableast` has minimal changes - public interface untouched, `from_ast` mapping method updated.

File: src/compiler/mod.rs [hunk 1]
```diff
@@ -2138,8 +2138,8 @@ fn compile_ast_function_call(
             Some(ast::Type::String) => Type::String,
             Some(ast::Type::PublicKey) => Type::PublicKey,
             Some(ast::Type::Bytes) => Type::Bytes,
-            Some(ast::Type::ForeignRecord { collection }) => Type::CollectionReference {
-                collection: collection.clone(),
+            Some(ast::Type::ForeignRecord { contract }) => Type::CollectionReference {
```
Comment (mateuszmlc) [inline src/compiler/mod.rs:2141]: With [#[serde(rename = ...)]](https://serde.rs/container-attrs.html) we could rename the other identifiers as well.
Although I'm not sure if we even need to support backwards compatibility, we have an `old-main-before-miden-compiler-merge` branch we could use for polybase-rust and if needed, selectively backport some changes to it in the future. @calummoore thoughts?

Comment (timmyjose) [reply src/compiler/mod.rs:2141]: @mateuszmlc  Just to be on the same page, you mean something like so:

```rust
Some(ast::Type::ForeignRecord { contract }) => Type::ContractReference {
        contract: contract.clone(),
    },
```

with `Type` being modified like so:

```rust
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub enum Type {
    ...

    #[serde(rename = "CollectionReference")]
    ContractReference {
        #[serde(rename = "collection")]
        contract: String,
    },
    ...
}

```

Is that right?

Comment (mateuszmlc) [reply src/compiler/mod.rs:2141]: Yes, exactly

**Review by calummoore — APPROVED** [review]

**Review by calummoore — APPROVED** [review]

**Comment by calummoore** [comment]
If I search "Collection" there are a bunch of other references to collection - especially in the AST create. Can we update these to?

@mateuszmlc I think we should also update CollectionReference to Contract reference?

**Comment by mateuszmlc** [comment]
> If I search "Collection" there are a bunch of other references to collection - especially in the AST create. Can we update these to?
> 
> @mateuszmlc I think we should also update CollectionReference to Contract reference?

I think that was for backwards compatibility, see the review comment above https://github.com/polybase/polylang/pull/114#discussion_r1331503073

**Comment by timmyjose** [comment]
I'll make changes, working through all the crates in `Polylang`, and reopen the P.R. for review when ready.

File: src/compiler/mod.rs (hunk 1 above)
Comment (timmyjose) [reply src/compiler/mod.rs]: Implemented and then removed the `rename` in the latest commit (since we're not looking at backward compatibility right now).

**Comment by timmyjose** [comment]
**All** references to `collection` have been replaced by `contract`. Only a couple of tests using `collection` have been kept in the main `Polylang` tests (tests`call_public_collection` and `call_any_call_collection`) - this is for testing that aliasing still works in `Polylang`. 

Given that we don't have backwards compatibily in mind, should we remove `collection` from the grammar itself (i.e., remove aliasing and simply use`contract`?)

(Also, some references in the `README` to `testnet` as well as associated unit tests and the `pull-collection` scripts are planned to  be removed as part of ENG-1086.)

**Comment by timmyjose** [comment]
Reopening for full review.

**Review by mateuszmlc — APPROVED** [review]

**Pull request merged**